pfsimulator/parflow_lib/pf_cudaloops.h

```diff
@@ -0,0 +1,1174 @@
+/**********************************************************************
+ *
+ *  Please read the LICENSE file for the GNU Lesser General Public License.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License (as published
+ *  by the Free Software Foundation) version 2.1 dated February 1999.
+ *
+ *  This program is distributed in the hope that it will be useful, but
+ *  WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms
+ *  and conditions of the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  License along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
+ *  USA
+ ***********************************************************************/
+
+/* @file
+ * @brief Contains macros, functions, and structs for CUDA compute kernels.
+ */
+
+#ifndef PF_CUDALOOPS_H
+#define PF_CUDALOOPS_H
+
+/*--------------------------------------------------------------------------
+ * Include CUDA headers
+ *--------------------------------------------------------------------------*/
+#include "pf_cudamain.h"
+#include "pf_cudamalloc.h"
+
+extern "C++"{
+
+#include <tuple>
+#include "cub.cuh"
+
+/*--------------------------------------------------------------------------
+ * CUDA block size definitions
+ *--------------------------------------------------------------------------*/
+#define BLOCKSIZE_MAX 1024
```

Comment:
BLOCKSIZE_MAX is currently the largest blocksize ParFlow is using, but it is also the largest blocksize supported by any currently available NVIDIA GPU architecture. This can also differ between different architectures. It is used for informing the compiler about how many registers should be available for the GPU kernel during the compilation. Another option is to use "_--maxrregcount 64_" compiler flag (I initially did this), but our NVIDIA contact told me that the recommended way is to specify this kernel-by-kernel basis by "_launch_bounds__(BLOCKSIZE_MAX)_". 

What I do not personally understand is why setting "_launch_bounds__(BLOCKSIZE_MAX)_" is required in the first place. Why the compiler could not just set the register usage according to the maximum blocksize of the target architecture if not otherwise specified? I asked this from our NVIDIA contact but he did not know the answer. Anyway, if the maximum number of available registers is not explicitly specified through either one of these mechanisms, the program crashes. 

The default 3D block dimensions in Parflow:
#define BLOCKSIZE_X 32
#define BLOCKSIZE_Y 8
#define BLOCKSIZE_Z 4

The blocksize for x dimension is set to 32, because the warp size for the current NVIDIA architectures is 32. Therefore, letting each thread in a warp access consecutive memory locations along the x-dimension results in best memory coalescence. Blocksizes along y and z-dimensions are less important. It is also important that the total blocksize (the product of x, y, and z-blocksizes) is divisible by the warp size.

Reply:
Please paste this into the header to help document these parameters...we have too much undocumented stuff in PF so need to write this valuable info down.

Reply:
done